stdlib/public/core/EmbeddedRuntime.swift

```diff
@@ -0,0 +1,132 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftShims
+
+public struct ClassMetadata {
+  var superclassMetadata: UnsafeMutablePointer<ClassMetadata>?
+
+  // There is no way to express the actual calling convention on the heap desroy
+  // function (swiftcc with 'self') currently, so let's use UnsafeRawPointer
+  // and a helper function in C (_swift_runtime_invoke_heap_object_destroy).
+  var destroy: UnsafeRawPointer
+}
+
+public struct HeapObject {
+  var metadata: UnsafeMutablePointer<ClassMetadata>
+
+  // TODO: This is just an initial support for strong refcounting only. We need
+  // to think about supporting (or banning) weak and/or unowned references.
+  var refcount: Int
+
+  static let immortalRefCount = -1
+}
+
+func alignedAlloc(size: Int, alignment: Int) -> UnsafeMutableRawPointer? {
+  let alignment = max(alignment, MemoryLayout<UnsafeRawPointer>.size)
+  var r: UnsafeMutableRawPointer? = nil
+  _ = posix_memalign(&r, alignment, size)
+  return r
+}
+
+@_cdecl("swift_slowAlloc")
+public func swift_slowAlloc(_ size: Int, _ alignMask: Int) -> UnsafeMutableRawPointer? {
+  let alignment: Int
+  if alignMask == -1 {
+    alignment = _swift_MinAllocationAlignment
+  } else {
+    alignment = alignMask + 1
+  }
+  return alignedAlloc(size: size, alignment: alignment)
+}
+
+@_cdecl("swift_slowDealloc")
+public func swift_slowDealloc(_ ptr: UnsafeMutableRawPointer?, _ size: Int, _ alignMask: Int) {
```

Comment:
Are you going to add the DO_NOT_FREE_BIT here, too?

Reply:
I don't think it should go here...? I thought swift_slowDealloc is not used on refcounted objects, and only used to implement things like UnsafePointer.deallocate(). This matches how swift_slowDealloc is implemented in the main runtime (no checking of anything, just a direct call to free).

Reply:
Ah sorry, I meant to add this comment for `swift_deallocClassInstance`

Reply:
Aha, then yes -- let me do that as a follow-up in <https://github.com/apple/swift/pull/68754/files>, I'll rebase it on top of this Swift runtime.